Sorting selective queries blog, plus plotly-cartesian 3.0.1

Alex-Monahan opened the pull request:
Here is an initial draft of the approximate sorting for fast selective queries blog!

Open to feedback!

Alex-Monahan commented:
And here is a PDF printout! 
[Approximate Sorting for Fast Selective Queries – DuckDB.pdf](https://github.com/user-attachments/files/20088181/Approximate.Sorting.for.Fast.Selective.Queries.DuckDB.pdf)

szarnyasg commented on _posts/2025-05-09-sorting-for-fast-selective-queries.md:
```diff
@@ -0,0 +1,660 @@
+CREATE OR REPLACE FUNCTION main.zip_varchar(i, j, num_chars := 6) AS (
+    -- By default using 6 characters from each string so that
+    -- if data is ASCII, we can fit it all in the 12 byte inline portion of DuckDB's string representation
+    [
```
I rewrote this with function chaining to make it more readable. Please double-check!

Alex-Monahan replied:
Super close! Just one extra space in the `array_to_string(' ')`. Thanks for the change - it is cleaner!

szarnyasg commented on _posts/2025-05-09-sorting-for-fast-selective-queries.md:
```diff
@@ -0,0 +1,660 @@
+In the rest of the post, we'll run different benchmarks to demonstrate how pruning can help reducing the amount of network traffic required to answer certain queries.
+
+Our hypothetical use case is to serve a dashboard where people can explore a single `origin`, a single `dest` (destination), or an `origin`–`dest` pair.
+The `origin` and `dest` columns are three-letter airport codes (like `LAX` or `JFK`).
```
I checked the dataset: airport codes are 3 letters, they cannot contain numbers.

Alex-Monahan replied:
Perfect, thank you!

szarnyasg commented on _posts/2025-05-09-sorting-for-fast-selective-queries.md:
```diff
@@ -0,0 +1,660 @@
+More details are included in the [Appendix](#appendix-experiment-details)!
+
+These plots display query runtime when pulling from a DuckDB file hosted on S3.
+Querying a local DuckDB file showed very similar patterns, but with much faster query times overall!
```
We may want to skip this. IMHO the key point of the techniques in the blog post is speeding up remote access.

Alex-Monahan replied:
I think I will reword it a bit to say that remote file access makes this a larger optimization. I think that will balance it!

Alex-Monahan replied:
Happy to change back, but I took a stab at rewording!

szarnyasg commented on _posts/2025-05-09-sorting-for-fast-selective-queries.md:
```diff
@@ -0,0 +1,660 @@
+We then use our most effective multi-column approach and sort by Hilbert encoding of `origin` and `dest` next.
+
+For each of the query patterns tested previously (filters on `origin`, `destination`, and `origin` / `destination`), we filter on 3 time ranges: the latest 1 week, 13 weeks, and 52 weeks.
+This yields a total of 9 scenarios.
```
It's a total of 3x3x3 = 27 scenarios. Part of the separation happens among the plots, another part happens within the plots. Please elaborate on this.

Alex-Monahan replied:
Good catch! Yes, the experiment needs some more explanation.

Alex-Monahan replied:
This will be in the follow up post! Thanks for the feedback!

szarnyasg commented on _posts/2025-05-09-sorting-for-fast-selective-queries.md:
```diff
@@ -0,0 +1,660 @@
+The `origin` and `dest` columns are three-letter airport codes (like `LAX` or `JFK`).
+We assume once you have picked one of those filters, you want all the rows that match and a few arbitrary columns.
+
+Our second experiment will add in a time component and additionally filter to the latest 1, 13, or 52 weeks of data in the dataset.
```
This is the first occurrence of the word `experiment`. What is the first experiment?

Alex-Monahan replied:
Ah, it was meant to be the first "hypothetical use case" of the dashboard. I will align the terminology!

szarnyasg commented on _posts/2025-05-09-sorting-for-fast-selective-queries.md:
```diff
@@ -0,0 +1,660 @@
+They operate on integers or floats, so a SQL macro is used to convert the first few characters of a `VARCHAR` column into an integer as a pre-processing step.
+More details are included in the [Appendix](#appendix-experiment-details)!
+
+These plots display query runtime when pulling from a DuckDB file hosted on S3.
```
What are the queries? How many are there? How were they run – once per queries, cold run? warm run? What DuckDB client was used (Python/CLI)?

Alex-Monahan replied:
I added those details - great call out. Thank you!

szarnyasg commented on _posts/2025-05-09-sorting-for-fast-selective-queries.md:
```diff
@@ -0,0 +1,660 @@
+The `rowgroup_id_count` column is a measurement of how many distinct row groups that a specific column value is present in, so it is an indicator of how much work DuckDB would need to do to pull all data associated with that value.
+
+> This calculation uses the [pseudo-column `rowid`]({% link docs/preview/sql/statements/select.md %}#row-ids), and it requires data to have been inserted in a single batch to be perfectly accurate.
+> It is directionally correct for data inserted in batches.
```
What does "directionally correct" mean?

Alex-Monahan replied:
Directionally correct is another way to say "approximate". I'm happy to change it to say approximate if that is clearer!

Alex-Monahan replied:
Made that change!

Alex-Monahan commented:
@szarnyasg , I've split things apart, and this includes just the first post in the series! I added some content around rules of thumb and previewed the follow up post a little bit as well. I also added the DuckDB file format diagrams (otherwise there would be no images at all! I thought they were helpful to the flow also).

szarnyasg commented:
Great! Let's see how it looks

Alex-Monahan commented:
> Great! Let's see how it looks

Thanks for the improvements!!
